### PR TITLE
Add Unsafe Analyzer using cargo-geiger

### PR DESCRIPTION
**Pull request by bitboom**
This PR introduces unsafe-analyzer to scan unsafe code using modified cargo-geiger.

This PR detects the Unsafe Blocks containing Dereference Operations .
Out of a total of 204 blocks, 40 were found to contain such operations automatically.

There are two types of Dereference Operations: (manually review)
- Dereference of raw pointer
- Access to union field

## How to run
```sh
$ ./scripts/unsafe-analyzer
[!] Adding hint: /home/sangwan/islet/lib/armv9a...
[!] Adding hint: /home/sangwan/islet/plat/fvp...
[!] Adding hint: /home/sangwan/islet/rmm...
[!] Adding hint: /home/sangwan/islet/lib/uart...
[!] Adding hint: /home/sangwan/islet/lib/vmsa...
[!] Running cargo geiger...
[!] Deleting hint: /home/sangwan/islet/lib/armv9a...
[!] Deleting hint: /home/sangwan/islet/plat/fvp...
[!] Deleting hint: /home/sangwan/islet/rmm...
[!] Deleting hint: /home/sangwan/islet/lib/uart...
[!] Deleting hint: /home/sangwan/islet/lib/vmsa...
[!] Reporting...
Total           | 204
Deref Operation | 40
Unclassified    | 163
Skipped         | 1
Exprs           | 749
Stmts           | 300
```

## Log (Dereference Operation)
```
$ cat out/unsafe-result.log | grep Deref
pub unsafe fn parse < 'a > (addr : usize) -> & 'a Self { let _crate_rmm : u8 ; & * (addr as * const Self) } ~ Method ~ 4 ~ 2 ~ Dereference Operation
pub unsafe fn parse_mut < 'a > (addr : usize) -> & 'a mut Self { let _crate_rmm : u8 ; & mut * (addr as * mut Self) } ~ Method ~ 4 ~ 2 ~ Dereference Operation
pub unsafe fn set_gpr (& mut self , idx : usize , val : u64) -> Result < () , Error > { let _crate_rmm : u8 ; if idx >= HOST_CALL_NR_GPRS { error ! ("out of index: {}" , idx) ; return Err (Error :: RmiErrorInput) ; } (* self . inner . val) . gprs [idx] = val ; Ok (()) } ~ Method ~ 14 ~ 4 ~ Dereference Operation
unsafe { let _crate_rmm : u8 ; & * self . inner . ptr } ~ Inner ~ 4 ~ 2 ~ Dereference Operation
unsafe { let _crate_rmm : u8 ; & * self . inner . ptr } ~ Inner ~ 4 ~ 2 ~ Dereference Operation
unsafe { let _crate_rmm : u8 ; & * self . inner . ptr } ~ Inner ~ 4 ~ 2 ~ Dereference Operation
unsafe { let _crate_rmm : u8 ; & * self . inner . ptr } ~ Inner ~ 4 ~ 2 ~ Dereference Operation
unsafe { let _crate_rmm : u8 ; & mut * self . inner . ptr } ~ Inner ~ 4 ~ 2 ~ Dereference Operation
unsafe { let _crate_rmm : u8 ; & mut * (addr as * mut T) } ~ Inner ~ 4 ~ 2 ~ Dereference Operation
unsafe { let _crate_rmm : u8 ; & * (addr as * const T) } ~ Inner ~ 4 ~ 2 ~ Dereference Operation
unsafe { let _crate_rmm : u8 ; & mut * PageTable :: < GuestPhysAddr , L2Table , entry :: Entry , { < L2Table as Level > :: NUM_ENTRIES } , > :: new_with_base (rtt_base) . unwrap () } ~ Inner ~ 4 ~ 2 ~ Dereference Operation
unsafe { let _crate_rmm : u8 ; & mut * PageTable :: < GuestPhysAddr , L0Table , entry :: Entry , { < L0Table as Level > :: NUM_ENTRIES } , > :: new_with_base (rtt_base) . unwrap () } ~ Inner ~ 4 ~ 2 ~ Dereference Operation
pub unsafe fn current () -> Option < & 'static mut VCPU < crate :: realm :: context :: Context > > { let _crate_rmm : u8 ; match TPIDR_EL2 . get () { 0 => None , current => Some (& mut * (current as * mut VCPU < crate :: realm :: context :: Context >)) , } } ~ Function ~ 7 ~ 2 ~ Dereference Operation
# [allow (dead_code)] pub unsafe fn init (config_addr : usize , ipa_width : usize) { let _crate_rmm : u8 ; let config : & mut RealmConfig = & mut * (config_addr as * mut RealmConfig) ; config . ipa_width = ipa_width ; } ~ Method ~ 6 ~ 3 ~ Dereference Operation
unsafe { let _crate_rmm : u8 ; & mut * RootPageTable :: < VirtAddr , L1Table , Entry , { < L1Table as Level > :: NUM_ENTRIES } > :: new_with_align (NUM_ROOT_PAGE , ALIGN_ROOT_PAGE ,) . unwrap () } ~ Inner ~ 4 ~ 2 ~ Dereference Operation
pub unsafe fn exit_gic_lrs_mut (& mut self) -> & mut [u64 ; 16] { let _crate_rmm : u8 ; & mut (* (* self . exit . inner) . gicv3 . inner) . lrs } ~ Method ~ 10 ~ 2 ~ Dereference Operation
pub unsafe fn set_exit_reason (& mut self , exit_reason : u8) { let _crate_rmm : u8 ; (* self . exit . inner) . exit_reason . val = exit_reason ; } ~ Method ~ 7 ~ 2 ~ Dereference Operation
pub unsafe fn set_esr (& mut self , esr : u64) { let _crate_rmm : u8 ; (* (* self . exit . inner) . sys_regs . inner) . esr = esr ; } ~ Method ~ 10 ~ 2 ~ Dereference Operation
pub unsafe fn set_far (& mut self , far : u64) { let _crate_rmm : u8 ; (* (* self . exit . inner) . sys_regs . inner) . far = far ; } ~ Method ~ 10 ~ 2 ~ Dereference Operation
pub unsafe fn set_hpfar (& mut self , hpfar : u64) { let _crate_rmm : u8 ; (* (* self . exit . inner) . sys_regs . inner) . hpfar = hpfar ; } ~ Method ~ 10 ~ 2 ~ Dereference Operation
pub unsafe fn set_gpr (& mut self , idx : usize , val : u64) -> Result < () , Error > { let _crate_rmm : u8 ; if idx >= NR_GPRS { error ! ("out of index: {}" , idx) ; return Err (Error :: RmiErrorInput) ; } (* self . exit . inner) . gprs . val [idx] = val ; Ok (()) } ~ Method ~ 15 ~ 4 ~ Dereference Operation
pub unsafe fn set_ripas (& mut self , base : u64 , size : u64 , state : u8) { let _crate_rmm : u8 ; (* (* self . exit . inner) . ripas . inner) . base = base ; (* (* self . exit . inner) . ripas . inner) . size = size ; (* (* self . exit . inner) . ripas . inner) . value = state ; } ~ Method ~ 30 ~ 4 ~ Dereference Operation
pub unsafe fn set_gic_lrs (& mut self , src : & [u64] , len : usize) { let _crate_rmm : u8 ; (* (* self . exit . inner) . gicv3 . inner) . lrs [.. len] . copy_from_slice (& src [.. len]) } ~ Method ~ 15 ~ 2 ~ Dereference Operation
pub unsafe fn set_gic_misr (& mut self , val : u64) { let _crate_rmm : u8 ; (* (* self . exit . inner) . gicv3 . inner) . misr = val ; } ~ Method ~ 10 ~ 2 ~ Dereference Operation
pub unsafe fn set_gic_vmcr (& mut self , val : u64) { let _crate_rmm : u8 ; (* (* self . exit . inner) . gicv3 . inner) . vmcr = val ; } ~ Method ~ 10 ~ 2 ~ Dereference Operation
pub unsafe fn set_gic_hcr (& mut self , val : u64) { let _crate_rmm : u8 ; (* (* self . exit . inner) . gicv3 . inner) . hcr = val ; } ~ Method ~ 10 ~ 2 ~ Dereference Operation
pub unsafe fn set_cntv_ctl (& mut self , val : u64) { let _crate_rmm : u8 ; (* (* self . exit . inner) . cnt . inner) . v_ctl = val ; } ~ Method ~ 10 ~ 2 ~ Dereference Operation
pub unsafe fn set_cntv_cval (& mut self , val : u64) { let _crate_rmm : u8 ; (* (* self . exit . inner) . cnt . inner) . v_cval = val ; } ~ Method ~ 10 ~ 2 ~ Dereference Operation
pub unsafe fn set_cntp_ctl (& mut self , val : u64) { let _crate_rmm : u8 ; (* (* self . exit . inner) . cnt . inner) . p_ctl = val ; } ~ Method ~ 10 ~ 2 ~ Dereference Operation
pub unsafe fn set_cntp_cval (& mut self , val : u64) { let _crate_rmm : u8 ; (* (* self . exit . inner) . cnt . inner) . p_cval = val ; } ~ Method ~ 10 ~ 2 ~ Dereference Operation
unsafe { let _crate_rmm : u8 ; & * (owner as * const Rd) } ~ Inner ~ 4 ~ 2 ~ Dereference Operation
unsafe { let _crate_vmsa : u8 ; let arr : [E ; N] = core :: array :: from_fn (| _ | E :: new ()) ; (* table) . entries = arr ; } ~ Inner ~ 7 ~ 3 ~ Dereference Operation
unsafe { let _crate_vmsa : u8 ; let arr : [E ; N] = core :: array :: from_fn (| _ | E :: new ()) ; (* table) . entries = arr ; } ~ Inner ~ 7 ~ 3 ~ Dereference Operation
unsafe { let _crate_vmsa : u8 ; let arr : [E ; L :: NextLevel :: NUM_ENTRIES] = core :: array :: from_fn (| _ | E :: new ()) ; (* subtable_ptr) . entries = arr ; } ~ Inner ~ 7 ~ 3 ~ Dereference Operation
unsafe { let _crate_vmsa : u8 ; & mut * (subtable_addr . as_usize () as * mut PageTable < A , L :: NextLevel , E , N >) } ~ Inner ~ 5 ~ 2 ~ Dereference Operation
unsafe { let _crate_vmsa : u8 ; & mut * (table_addr as * mut PageTable < A , L :: NextLevel , E , N >) } ~ Inner ~ 4 ~ 2 ~ Dereference Operation
unsafe { let _crate_vmsa : u8 ; & mut * (table_addr as * mut PageTable < A , L :: NextLevel , E , N >) } ~ Inner ~ 4 ~ 2 ~ Dereference Operation
unsafe { let _crate_vmsa : u8 ; & * (self . addr as * const T) } ~ Inner ~ 5 ~ 2 ~ Dereference Operation
unsafe { let _crate_vmsa : u8 ; & mut * (self . addr as * mut T) } ~ Inner ~ 5 ~ 2 ~ Dereference Operation
```

### CODE REVIEW

**Review by jinbpark — APPROVED** [review]

**Review by zpzigi754 — APPROVED** [review]
Really nice to customize `cargo-geiger` for categorizing unsafe usages!

**Review by nook1208 — APPROVED** [review]

File: scripts/pylint.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+set -e
+
+ROOT=$(git rev-parse --show-toplevel)
+HERE=$ROOT/scripts
+
+# TODO: Add fvp-cca
+cp $HERE/unsafe-analyzer $HERE/unsafe-analyzer.py
```
Comment (zpzigi754) [inline scripts/pylint.sh:9]: Is this copy unavoidable? I mean `pylint` cannot be applied to the file without py extension?

Comment (bitboom) [reply scripts/pylint.sh]: I wasn't able to do it in my environment. However, after trying again, it worked without any problems. I think there was a mistake on my part. Thank you for pointing out :)

**Comment by jinbpark** [comment]
According to the attached log, the majority of it comes from "register access" which is inevitable in system software.
It'd be interesting to think about the way we attach "safety tag" to them.

I think, we basically have two options--
(1) might be good enough. (2) might require some work.

(1) treat all register accesses as safe operations, without any restriction
```
pub unsafe fn set_esr(&mut self, esr: u64) {
        // Safety: ....
        (*(*self.exit.inner).sys_regs.inner).esr = esr;
    }
```

(2) make some restrictions if needed for safety
```
// e.g., enforce that a safety tag is attached only if a specific bit of a hardware register is disabled or enabled.
```

**Review by hihi-wang — APPROVED** [review]

**Pull request merged**